File: Second_Semester/Discrete-Structure/Recursive-Problems/BinarySearch.c
#include <stdio.h>

int binarySearch(int arr[], int low, int high, int key) {
    if (low <= high) {
        int mid = low + (high - low) / 2;
        if (arr[mid] == key)
            return mid;
        else if (arr[mid] > key)
            return binarySearch(arr, low, mid - 1, key);
        else
            return binarySearch(arr, mid + 1, high, key);
    }
    return -1;
}

int main() {
    int arr[] = { 2, 4, 6, 8, 10, 12, 14, 16, 18, 20 };
    int n = sizeof(arr)/sizeof(arr[0]);
    int key;
    printf("Enter key to search: ");
    scanf("%d",&key);
    int index = binarySearch(arr, 0, n - 1, key);
    if (index == -1)
        printf("%d not found in the array", key);
    else
        printf("%d found at index %d", key, index);
    return 0;
}

File: Third_Semester/Data-Structure-And-Algorithms-with-JAVA/Searching/BinarySearch.java
import java.util.*;

public class BinarySearch {
    public int binarySearch(int arr[], int key) {
        int low = 0;
        int high = arr.length - 1;
        while (low <= high) {
            int mid = low + (high - low) / 2;

            if (arr[mid] == key) {
                return mid;
            }

            if (key > arr[mid]) { // if key is greater, ignore the left half
                low = mid + 1;
            } else {
                high = mid - 1;
            }

        }
        return -1;

    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        BinarySearch search = new BinarySearch();

        // Array should be Sorted for binary search
        int arr[] = { 1, 3, 5, 8, 56, 76, 82, 99 };
        System.out.println("Enter a key to search:");
        int key = sc.nextInt();

        int res = search.binarySearch(arr, key);

        if (res == -1) {
            System.out.println("Sorry! key not found");
        } else {
            System.out.println("Key found at index " + res);
        }
    }

}
